Jint/Pooling/EnginePool.cs

```diff
@@ -0,0 +1,110 @@
+using System;
+using Esprima;
+using Esprima.Ast;
+using Jint.Native;
+
+namespace Jint.Pooling
+{
+    /// <summary>
+    /// Offers a pool of engines that can be initialized to certain state and then used with temporary state that
+    /// will be discarded.
+    /// </summary>
+    public sealed class EnginePool
+    {
+        private readonly ObjectPool<Engine> _enginePool;
+
+        public EnginePool(int size, Options options, Action<IPooledEngine> initialize)
+        {
+            _enginePool = new ObjectPool<Engine>(factory: () =>
+            {
+                var engine = new Engine(options);
+                initialize?.Invoke(new EngineWrapper(engine));
+                return engine;
+            }, size);
+        }
+
+        public void ExecuteInIsolatedContext(Action<IPooledEngine> action)
+        {
+            Engine engine;
+            lock (_enginePool)
+            {
+                engine = _enginePool.Allocate();
+            }
+
+            using (engine.EnterIsolatedContext())
```

Comment:
With the interface, the pool could return an IEngine directly, which would be the result of this call.